### PR TITLE
Custom debounce time for publisher

### PR DESCRIPTION
**Pull request by Medvedu**
A minor improvement which is allow to set custom debounce time for publisher.

### CODE REVIEW

File: lib/table_sync/publisher.rb [hunk 1]
```diff
@@ -4,26 +4,27 @@ class TableSync::Publisher < TableSync::BasePublisher
   DEBOUNCE_TIME = 1.minute
 
   # 'original_attributes' are not published, they are used to resolve the routing key
-  def initialize(object_class, original_attributes, destroyed: nil, confirm: true, state: :updated)
+  def initialize(object_class, original_attributes, **opts)
     @object_class = object_class.constantize
     @original_attributes = filter_safe_for_serialization(original_attributes.deep_symbolize_keys)
-    @confirm = confirm
+    @confirm = opts[:confirm].nil? ? true : opts[:confirm]
```
Comment (akxcv) [inline lib/table_sync/publisher.rb]:
```suggestion
    @confirm = opts.fetch(:confirm, true)
```

File: lib/table_sync/publisher.rb [hunk 2]
```diff
@@ -4,26 +4,27 @@ class TableSync::Publisher < TableSync::BasePublisher
...
+    @debounce_time = opts[:debounce_time].nil? ? nil : opts[:debounce_time].seconds
```
Comment (akxcv) [inline lib/table_sync/publisher.rb]:
```suggestion
    @debounce_time = opts[:debounce_time]&.seconds
```

File: lib/table_sync/publisher.rb [hunk 3]
```diff
@@ -4,26 +4,27 @@ class TableSync::Publisher < TableSync::BasePublisher
...
   def publish
-    return enqueue_job if destroyed?
+    return enqueue_job if destroyed? || debounce_time_zero?
 
-    sync_time = Rails.cache.read(cache_key) || current_time - DEBOUNCE_TIME - 1.second
+    sync_time = Rails.cache.read(cache_key) || current_time - debounce_time_offset - 1.second
     return if sync_time > current_time
 
-    next_sync_time = sync_time + DEBOUNCE_TIME
+    next_sync_time = sync_time + debounce_time_offset
```
Comment (FarnoUmb) [inline lib/table_sync/publisher.rb]: Maybe add 1 second that were subtracted in `sync_time `?

File: lib/table_sync/orm_adapter/sequel.rb [hunk 4]
```diff
@@ -19,13 +19,29 @@ def attributes(object)
...
       unless_predicate = to_predicate(opts.delete(:unless), false)
-      raise "Only :if and :unless options are currently supported for Sequel hooks" if opts.any?
+      debounce_time    = opts.delete(:debounce_time) { |key| opts.fetch(key, nil) }
 
+      if opts.any?
+        raise "Only :if, :skip_debounce and :unless options are currently " \
+              "supported for Sequel hooks"
+      end
+
+      _setup_sync(klass, if_predicate, unless_predicate, debounce_time)
```
[the remark this answers is not in the document]
Comment (Medvedu) [reply lib/table_sync/orm_adapter/sequel.rb]: Agree with you, however rubocop said "method has too many lines" so i decided to separate setup_sync onto two methods.

File: lib/table_sync/publisher.rb (hunk 1 above)
Comment (Medvedu) [reply lib/table_sync/publisher.rb]: alright, thanks

File: lib/table_sync/publisher.rb (hunk 3 above)
Comment (Medvedu) [reply lib/table_sync/publisher.rb]: Should we add extra second? I've check original code and didn't found any extra calculation there.

File: lib/table_sync/publisher.rb (hunk 2 above)
Comment (Medvedu) [reply lib/table_sync/publisher.rb]: not actual anymore

File: lib/table_sync/orm_adapter/sequel.rb [hunk 5]
```diff
@@ -19,13 +19,29 @@ def attributes(object)
...
+      registrate_callbacks(klass, if_predicate, unless_predicate, debounce_time)
+    end
+
+    def to_predicate(val, default)
+      return val.to_proc if val.respond_to?(:to_proc)
+
+      -> (*) { default }
+    end
+
+    def registrate_callbacks(klass, if_predicate, unless_predicate, debounce_time)
```
Comment (akxcv) [inline lib/table_sync/orm_adapter/sequel.rb]: `register`

File: docs/synopsis.md [hunk 6]
```diff
@@ -82,8 +82,8 @@ to default Rabbit gem configuration).
 
 # Manual publishing
 
-`TableSync::Publisher.new(object_class, original_attributes, confirm: true, state: :updated)` where
-state is one of `:created / :updated / :destroyed` and `confirm` is Rabbit's confirm delivery flag
+`TableSync::Publisher.new(object_class, original_attributes, confirm: true, state: :updated, debounce_time: 45)` 
+where state is one of `:created / :updated / :destroyed` and `confirm` is Rabbit's confirm delivery flag and optional param `debounce_time` determines debounce time in seconds.
```
Comment (enthrops) [inline docs/synopsis.md]: please specify that default is 60

File: lib/table_sync/orm_adapter/sequel.rb [hunk 7]
```diff
@@ -19,13 +19,29 @@ def attributes(object)
     def setup_sync(klass, **opts)
       if_predicate     = to_predicate(opts.delete(:if), true)
       unless_predicate = to_predicate(opts.delete(:unless), false)
-      raise "Only :if and :unless options are currently supported for Sequel hooks" if opts.any?
+      debounce_time    = opts.delete(:debounce_time) { |key| opts.fetch(key, nil) }
```
Comment (enthrops) [inline lib/table_sync/orm_adapter/sequel.rb]: I think `opts.delete(:debounce_time)` should be sufficient here

File: lib/table_sync/orm_adapter/sequel.rb [hunk 8]
```diff
@@ -19,13 +19,29 @@ def attributes(object)
...
 
+      if opts.any?
+        raise "Only :if, :skip_debounce and :unless options are currently " \
+              "supported for Sequel hooks"
+      end
+
+      registrate_callbacks(klass, if_predicate, unless_predicate, debounce_time)
```
Comment (enthrops) [inline lib/table_sync/orm_adapter/sequel.rb]: register

File: lib/table_sync/orm_adapter/sequel.rb (hunk 7 above)
Comment (enthrops) [inline lib/table_sync/orm_adapter/sequel.rb]: Debounce time customization should also be supported for AR

File: lib/table_sync/publisher.rb [hunk 9]
```diff
@@ -4,26 +4,27 @@ class TableSync::Publisher < TableSync::BasePublisher
   DEBOUNCE_TIME = 1.minute
 
   # 'original_attributes' are not published, they are used to resolve the routing key
-  def initialize(object_class, original_attributes, destroyed: nil, confirm: true, state: :updated)
```
Comment (enthrops) [inline lib/table_sync/publisher.rb]: Publisher will now fail if state is not passed explicitly, let's use **opts, but set its values manually considering defaults

**Comment by akxcv** [comment]
Commit author data is wrong. Please force push with correct author.

File: lib/table_sync/publisher.rb [hunk 10]
```diff
@@ -4,26 +4,28 @@ class TableSync::Publisher < TableSync::BasePublisher
...
     @original_attributes = filter_safe_for_serialization(original_attributes.deep_symbolize_keys)
-    @confirm = confirm
+    @confirm = opts.fetch(:confirm, true)
+    @debounce_time = opts[:debounce_time]&.seconds || DEBOUNCE_TIME
 
-    if destroyed.nil?
-      @state = validate_state(state)
+    if opts[:destroyed].nil?
+      @state = opts.fetch(:state, :update)
+      validate_state
```
Comment (akxcv) [inline lib/table_sync/publisher.rb]: `state.to_sym` is now lost, might break smth

**Pull request closed without merging**